src/FastExpressionCompiler/FastExpressionCompiler.cs

```diff
@@ -3014,7 +3066,6 @@ private static bool TryEmitIndexAssign(IndexExpression indexExpr, Type instType,
                 var nestedNonPassedParams = nestedClosureInfo.NonPassedParameters;
                 if (nestedNonPassedParams.Length == 0)
                 {
-                    il.Emit(OpCodes.Castclass, nestedLambdaType);

```

Comment:
I removed this castclass too, because its done on above line 
`il.Emit(OpCodes.Ldelem_Ref); // load the array item object on stack and cast it lambda`